### PR TITLE
feat(store): add foundation for v2 data modelling

### PR DESCRIPTION
**Pull request by alvrs**
**TODO**
- [x] Streamline naming in Storage and Memory libraries (probably just use load and store instead of read and write?)
- [x] Move all internal functions to a separate StoreCoreInternal library to discourage using it
- [x] Change Storage library functions to make it clearer which argument is offset and which is length
- [x] Turn all storage pointer to uint256 for consistency (uint256 is better than bytes32 because it's easier to do arithmetic on)
- [x] change `bytes.concat` to `abi.encodePacked` where possible
- [x] Change all manual gas logs in tests to `// !gasreport` comments to use the `mud gas-report` cli command

## Follow ups
- [ ] Combine or further split `Bytes`, `Buffer`, `Memory` - they have some overlap and it’s not entirely clear where the boundaries are right now

**Low hanging fruits for gas improvements**

- [ ] Add `pushToDynamicField` and `popFromDynamicField` functions to store core lib to avoid writing entire array when adding a single element (eg `CallbackArray`)
- [ ] `Storage.write` and `Storage.read`
- [ ] Decoding values returned by store core library using as little memory copies as possible
- [ ] All low level libraries (`Bytes`, `Buffer`, `Memory`, `Storage`, `SchemaLib`, `PackedCounterLib`)
- [ ] Maybe: use abi.encodePacked for passing data into StoreCore? We'd need to construct the `PackedCounter` in StoreCore, but gas implications would be interesting. Maybe we can even use abi.encodePacked for data coming out of StoreCore, and have utils to decode them into non-packed with a given schema.

### CODE REVIEW

File: packages/solecs/v2/tables/Vec2Table.sol
Comment (holic) [inline packages/solecs/v2/tables/Vec2Table.sol:15]: would this need to be called `Schema` to be picked up by autogen or could this be `Vec2`? thinking that we'll probably be passing these kinds of data around and having everything named `Schema` would be annoying

File: packages/solecs/v2/tables/Vec2Table.sol
Comment (holic) [inline packages/solecs/v2/tables/Vec2Table.sol:18]: what would a scalar look like in this model? would we always have to pass around a single-value struct and call `.value` on it (like we do on the client)? is there a more ergonomic approach?

Comment (dk1a) [reply packages/solecs/v2/tables/Vec2Table.sol:18]: I think it's entirely possibly to make a special case for scalars, which would be entirely autogen's responsibility.
`Schema` is more of a convenience thing for users - StoreCore doesn't care about it.

File: packages/solecs/v2/tables/Vec2Table.sol
Comment (alvrs) [reply packages/solecs/v2/tables/Vec2Table.sol:15]: Agree, it should a named struct for each table instead of `Schema` everywhere!

Comment (dk1a) [reply packages/solecs/v2/tables/Vec2Table.sol:15]: For autogen I was imagining a sol-type-file/ts/json which is the config with all the schemas, from which all the .sol files are spawned

File: packages/solecs/v2/tables/Vector2Table.sol
Comment (holic) [inline packages/solecs/v2/tables/Vector2Table.sol:13]:
```suggestion
bytes32 constant id = keccak256("mud.store.table.vector2");
```

**Comment by alvrs** [comment]
very much WIP - don't look

**Comment by dk1a** [comment]
a tiny improvement suggestion:
`abi.encodePacked(x, y);` vs `bytes.concat(bytes4(x), bytes4(y));` etc - shorter and no need for typecasts

File: packages/store/src/StoreCore.sol
Comment (holic) [inline packages/store/src/StoreCore.sol:34]: I noticed some of our error names use `_` and recently others do not. We should prob converge on some pattern.

I also wonder about similarly prefixing errors with `Mud`. Seems like that's the direction folks are going with ERC721, etc. errors because it helps with overlapping names.

That said, I'm now seeing `event MudStore...(...)` in the context of `StoreCore` and wondering again about no prefix vs. prefixing everything internal. 🙈  Can't decide!

File: packages/store/src/IStore.sol
Comment (holic) [reply packages/store/src/IStore.sol]: second argument still needs `hooks` -> `hook` (I should have suggested a diff)

**Comment by alvrs** [comment]
Alright, not 100% done yet (see TODOs in PR description), but ready to have a closer look @dk1a @holic @ludns @authcall @Kooshaba. I know this is a massive PR, but let's try to get it to a point where we're happy with it as a foundation, and then iterate further in separate follow-up PRs (which makes it easier to work on follow up tasks in parallel). Also, once this is merged into `v2`, we can see gas impact of all follow up PRs via the diff in the `gas-report.txt` file.

File: packages/store/src/IStore.sol
Comment (alvrs) [reply packages/store/src/IStore.sol]: ahh good catch

**Comment by alvrs** [comment]
merging into `v2` branch now, so we can start working on followups in separate PRs

File: packages/store/src/Buffer.sol
Comment (dk1a) [inline packages/store/src/Buffer.sol:201]: You copy in 32 byte chunks, whereas the memory isn't word-aligned, so for if e.g. `_dataLength == 8`, memory corruption can happen.
[A verbose but pretty efficient alternative](https://gist.github.com/dk1a/6944af1557268e86b6d9d7ad801a93f9), for long bytes it can even costs less gas

File: packages/store/src/Buffer.sol
Comment (dk1a) [inline packages/store/src/Buffer.sol:11]: I wonder if 10/10/10 split for ptr/length/capacity would be better, then slicing can be done in-place without copying bytes (since you don't need a new length field in-memory)
[LinkedListLib](https://github.com/brockelmore/memmove/blob/master/src/LinkedList.sol) has a 3-way bitpacking example, although it's not very relevant

Comment (dk1a) [reply packages/store/src/Buffer.sol:11]: But also, could you get rid of capacity entirely? it seems like an intermediate internal concept, you really need it only for checking overflow. If optimization converges on appendUnchecked anyways, capacity might be useless?

File: packages/store/src/Buffer.sol
Comment (dk1a) [inline packages/store/src/Buffer.sol:250]: wdyt about using `bytes32[] memory array` instead of `uint256 arrayPtr`? Cast will also need to be changed, but otherwise everything including gas (just tested) stays the same, and we avoid using pointers more than necessary.
It'd also be clearer that `toArray` can't make `bytes[]`/`string[]` arrays

File: packages/store/src/Bytes.sol
Comment (dk1a) [inline packages/store/src/Bytes.sol:128]: seems easier to just use `abi.encodePacked` instead of having 96 converter methods; I guess this is a leftover after `bytes.concat`?

File: packages/store/src/Bytes.sol
Comment (dk1a) [inline packages/store/src/Bytes.sol:211]: If `Buffer` becomes 10/10/10 or ditches capacity, these methods can operate on `Buffer` directly

File: packages/store/src/Bytes.sol
Comment (dk1a) [inline packages/store/src/Bytes.sol:242]:
```suggestion
  function setLengthInPlace(bytes memory input, uint256 length) internal pure returns (bytes memory) {
    assert(length <= input.length);
```
changes cost from 16 to 82

File: packages/store/src/Bytes.sol
Comment (dk1a) [inline packages/store/src/Bytes.sol:705]:
```suggestion
```
unused

File: packages/store/src/PackedCounter.sol
Comment (dk1a) [inline packages/store/src/PackedCounter.sol:124]:
```suggestion
    // Update the total value according to the difference
    if (newValueAtIndex >= currentValueAtIndex) {
      accumulator += newValueAtIndex - currentValueAtIndex;
    } else {
      accumulator -= currentValueAtIndex - newValueAtIndex;
    }
```
I'd prefer uints when possible, and this saves some gas: 884 -> 830

File: packages/store/src/Schema.sol
Comment (dk1a) [inline packages/store/src/Schema.sol:63]:
```suggestion
    schema = Bytes.setBytes2(schema, 0, bytes2(length)); // 2 length bytes
```

File: packages/store/src/StoreCore.sol
Comment (dk1a) [inline packages/store/src/StoreCore.sol:220]: Deleting dynamic data seems unnecessary, since you delete its length; a new record that sets the new length would be responsible for initializing the elements

File: packages/store/src/StoreCore.sol
Comment (dk1a) [inline packages/store/src/StoreCore.sol:234]:
```suggestion
  /**
   * Get full data (static and dynamic) for the given table and key tuple, loading schema from storage
   */
```

File: packages/store/src/StoreCore.sol
Comment (dk1a) [inline packages/store/src/StoreCore.sol:263]: Coming back to [the Buffer comment](https://github.com/latticexyz/mud/pull/352/files#r1096575924) - here it could be something like:
```solidity
bytes memory result = new bytes(outputLength);
Buffer buffer = Buffer_.fromBytes(result);
```
I guess I'm suggesting a `Slice` at this point, not `Buffer`

File: packages/store/src/StoreCore.sol
Comment (dk1a) [inline packages/store/src/StoreCore.sol:400]:
```suggestion
  /************************************************************************
   *
   *    GET DATA
   *
   ************************************************************************/

  /**
   * Get full static record for the given table and key tuple (loading schema from storage)
   */
  function _getStaticData(bytes32 table, bytes32[] memory key) internal view returns (bytes memory) {
    Schema schema = _getSchema(table);
    return _getStaticData(table, key, schema);
  }
```

File: packages/store/src/StoreCore.sol
Comment (dk1a) [inline packages/store/src/StoreCore.sol:475]: the comment seems confusing, like it used to skip something but now doesn't

File: packages/store/src/System.sol
Comment (dk1a) [inline packages/store/src/System.sol:4]: https://github.com/OpenZeppelin/openzeppelin-contracts/blob/master/contracts/metatx/ERC2771Context.sol

File: packages/store/src/Types.sol
Comment (dk1a) [inline packages/store/src/Types.sol:23]: avoiding both linear and binary searches is probably a good idea, something like https://github.com/dk1a/mud/blob/c4cae23ac0284d6d475b683ad145daaf748e910e/packages/solecs/v2/SchemaType.sol#L225

File: packages/store/src/Utils.sol
Comment (dk1a) [inline packages/store/src/Utils.sol:20]: Since you changed it from bytes to bits the comments are confusing.
But it should probably be inlined into `_loadPartialWord` and `_storePartialWord` and combined with the other bit manipulation for efficiency anyways

**Review by dk1a — COMMENTED** [review]
IStore looks good; though I'm not sure how useful MudStoreDeleteRecord is, since it's basically like MudStoreSetRecord with all zeros. There doesn't seem to be a built-in `has` like with components (which makes sense, since this is generic storage and you usually don't waste a slot on an existence flag)

To summarize about `Buffer` - imo it currently combines the roles of a static buffer and slice. I'd explore not storing capacity, refactoring it to a `Slice`, and pairing it with `bytes` where necessary (and `bytes` length can be reduced if that's required).
Tried to avoid talking about gas for the most part.

Should I wait for merge before moving to gas optimizations and codegen?

File: packages/store/src/Buffer.sol
Comment (alvrs) [reply packages/store/src/Buffer.sol:201]: Good point! Mind changing this in a follow up PR?

File: packages/store/src/Buffer.sol
Comment (alvrs) [reply packages/store/src/Buffer.sol:11]: Yeah I agree - the whole point of this library is to make encoding/decoding data in `StoreCore` and the typed libs as cheap as possible, and I think it can be optimized to the use cases we have there instead of trying to be as general as possible. Since in the use cases we have there we always know the capacity in advance and just use the buffer as an intermediate to append to (or slice from), I think we can get rid of the explicit capacity. Mind changing this in a follow up?

File: packages/store/src/Buffer.sol
Comment (alvrs) [reply packages/store/src/Buffer.sol:250]: agree with this change!

File: packages/store/src/Bytes.sol
Comment (alvrs) [reply packages/store/src/Bytes.sol:211]: Agree!

File: packages/store/src/Bytes.sol
Comment (alvrs) [reply packages/store/src/Bytes.sol:242]: This started out as a purely internal function, but it seems not to be used anywhere anymore. Will remove.

File: packages/store/src/StoreCore.sol
Comment (alvrs) [reply packages/store/src/StoreCore.sol:263]: Agree with this!

File: packages/store/src/StoreCore.sol
Comment (alvrs) [reply packages/store/src/StoreCore.sol:475]: indeed a leftover! Removing

File: packages/store/src/System.sol
Comment (alvrs) [reply packages/store/src/System.sol:4]: yeah true, will remove the comment. For future reference - we don't need a trusted forwarder check here, because systems don't contain own storage, but instead act on the caller's storage (either via delegatecall or via callbacks, see `StoreSwitch`), so any caller can be "trusted".

File: packages/store/src/Utils.sol
Comment (alvrs) [reply packages/store/src/Utils.sol:20]: True, removed the comments for now, but agree with inlining it later

**Comment by alvrs** [comment]
First of all, thanks a lot for your review!

> though I'm not sure how useful MudStoreDeleteRecord is, since it's basically like MudStoreSetRecord with all zeros. There doesn't seem to be a built-in has like with components (which makes sense, since this is generic storage and you usually don't waste a slot on an existence flag)

There is a subtle difference: with support for callbacks via the `IStoreHook`, it's possible to create an "on-chain indexer" to store whether "an entity has a component", because the hooks for `setRecord` and `deleteRecord` are different. To make off-chain indexers consistent with this, they have to distinguish between `MudStoreSetRecord` and `MudStoreDeleteRecord`.

> To summarize about Buffer - imo it currently combines the roles of a static buffer and slice. I'd explore not storing capacity, refactoring it to a Slice, and pairing it with bytes where necessary (and bytes length can be reduced if that's required).

I agree with this, commented above!

> Should I wait for merge before moving to gas optimizations and codegen?

I just pushed fixes for the comments marked as resolved, and think we should work on the unresolved ones in follow up PRs.
Happy to merge into `v2` tomorrow if you agree, so you can get started on optimization and autogen.

**Comment by dk1a** [comment]
> There is a subtle difference: with support for callbacks via the IStoreHook, it's possible to create an "on-chain indexer" to store whether "an entity has a component", because the hooks for setRecord and deleteRecord are different. To make off-chain indexers consistent with this, they have to distinguish between MudStoreSetRecord and MudStoreDeleteRecord.

Right, makes sense. Later I'd been thinking it might have to do with indexers, hadn't thought about hooks.

> Mind changing this in a follow up PR?

Sure, especially with Buffer comments I was expecting to do the changes myself, it overlaps with gas stuff too.

> Happy to merge into v2 tomorrow if you agree, so you can get started on optimization and autogen.

lgtm

**Review by dk1a — APPROVED** [review]

**Pull request merged**
